### PR TITLE
Update to work with new website

### LINKED ISSUE #9: New Yosemite Planning Your Wilderness Permit page prevents automated access to the trailhead report
Currently, the trailhead report webpage that hackjohn uses at https://www.nps.gov/yose/planyourvisit/fulltrailheads.htm shows the following:

> image

The "improved version" at https://yosemite.org/planning-your-wilderness-permit/ places the trailhead report behind a capcha:

>  image

So it looks like users will no longer be able to perform automated checks of the trailhead report, like Hackjohn used to enable. This is a bit of a bummer, but perhaps Hackjohn was a victim of its own success!

If anyone is aware of a workaround or finds the trailhead available from a public API endpoint, let us know below and Hackjohn might be able to resurrect itself.

### PR DESCRIPTION
**Pull request by djcunningham0**
Fixes #9. The README has been updated to reflect all of the changes and enhancements described below.

# Core changes

These changes allow hackjohn to work with the new permit website. The major changes to the core code are:
* Use the 2Captcha service to bypass the Recaptcha.
* Use the API endpoints to retrieve permit and trailhead information. It's no longer an HTML table that can be easily scraped, so we have to call the API and then use the quota logic to calculate permit availability.

# Other enhancements

* Add support for Twilio SMS notifications.
* With the Recaptcha and API changes, it became helpful to modularize parts of the code. I ended up rewriting more of the code than I originally expected, but I think the result is more modular and hopefully easier to maintain. (I'm open to restructuring if it would make it easier to use or maintain.)

### CODE REVIEW

**Comment by djcunningham0** [comment]
The checks are going to fail now because I didn't put a valid 2Captcha API key into the code, so it's going to fail to solve the Recaptcha. We might be able to do something with [encrypted secrets](https://docs.github.com/en/actions/reference/encrypted-secrets) in the repo, but I'm unfamiliar with GitHub Actions in general.

File: config.py
Comment (dhimmel) [inline config.py:72]: ~~Do we want to put a comment that users will have to change `RECAPTCHA_SITE_KEY`?~~

nevermind, the RECAPTCHA_SITE_KEY is from the yosemite website.

File: hackjohn.py
Comment (dhimmel) [inline hackjohn.py]: Does it make sense to move this to config.py?

File: README.md
Comment (dhimmel) [inline README.md]: I tend to use one-phrase-per-line when tracking prose with git. It helps with diffs.

Is this a formatting style you can retain in the README?

**Review by dhimmel — COMMENTED** [review]
This is fantastic! A huge thanks for putting this together and sharing it.

hackjohn is reborn, smarter and stronger.

A couple small comments. Let me know what you think about them.

Feel free to acknowledge your contribution in the README or any other way you see fit. Would you also like to be a maintainer after the PR is merged?

**Comment by dhimmel** [comment]
> The checks are going to fail now because I didn't put a valid 2Captcha API key into the code, so it's going to fail to solve the Recaptcha. We might be able to do something with encrypted secrets in the repo, but I'm unfamiliar with GitHub Actions in general.

adding github actions secrets is quite easy. I can help with that. Probably we could allow setting the secrets by environment variables. No need to block this PR on getting CI working. Did you have extra credit on your 2captcha account that you'd be interested in donating? If not I could also look into setting that up?

File: config.py
Comment (djcunningham0) [reply config.py:72]: Right, that won't change unless they change the website. I stuck it in the config file so users won't think they need to modify it.

File: hackjohn.py
Comment (djcunningham0) [reply hackjohn.py]: This is what I intended when I set it up that way:
* `hackjohn.py` is the only file the user will need to edit before running. It contains all user API keys; user preferences around dates, trailheads, etc.; and notification settings.
* `config.py` contains constants that the user doesn't need to see or know about. We might need to change those values if the website or permit process changes in the future, so it's kind of helpful from a development to have those values in a separate file rather than hardcoding them into hackjohn.py.

I'm open to restructuring though. We could either move all of parameters to config.py, or we could add another file for those parameters (for example, user_config.py for parameters the user should edit and system_config.py for the values that shouldn't be changed). Or we could even go back to having everything included in hackjohn.py and not import any other files.

I don't really feel strongly any which way. Do you have any preference?

File: README.md
Comment (djcunningham0) [reply README.md]: Yes, I can do that. I usually try to stick to 80 characters per line for readability in the unrendered file, but I see the advantage of this for diffs. I'll reformat and commit shortly.

**Comment by djcunningham0** [comment]
> Feel free to acknowledge your contribution in the README or any other way you see fit. Would you also like to be a maintainer after the PR is merged?

Thanks, I'll drop a quick note in there somewhere, and I'd love to be a maintainer! Feel free to add me as a collaborator on the repo, or whatever you need to do.

**Comment by djcunningham0** [comment]
> Probably we could allow setting the secrets by environment variables. No need to block this PR on getting CI working. Did you have extra credit on your 2captcha account that you'd be interested in donating? If not I could also look into setting that up?

I have a couple dollars in a 2captcha account which should be more than enough. It'll cost less than a penny each time the test is run. Just let me know where I need to add the API key, or I can send it to you privately to add it.

**Comment by dhimmel** [comment]
> Feel free to add me as a collaborator on the repo

Sent a pending invite to as you as a collaborator. For this PR, wait for me to "Approve" on GitHub and then you can go ahead and merge it.

I don't think you'll be able to [add](https://docs.github.com/en/actions/reference/encrypted-secrets) the secret:

> To create secrets for a user account repository, you must be the repository owner.

I will add the secret at https://github.com/dhimmel/hackjohn/settings/secrets/actions/new

File: hackjohn.py
Comment (dhimmel) [reply hackjohn.py]: My worry with the parameters in hackjohn.py is that this file might be a bit more daunting to users compared to a more simple config file. So in that sense I like the idea of:

1. user_config.py / system_config.py
2. config.py with both but comments separating the two sections.

Up to you. Also can be done as a subsequent PR to this.

**Review by dhimmel — APPROVED** [review]
Thanks again for reviving hackjohn.

The secret and config changes can occur in subsequent PRs or this one (up to you).

Looks good to me.

File: hackjohn.py
Comment (djcunningham0) [reply hackjohn.py]: I'll keep the single config.py file and separate the two sections with comments. Should be quick so I'll do it in this PR.

**Comment by djcunningham0** [comment]
> You can email it

Just sent you an email. No rush on adding the secret to the repo -- I won't wait on that to merge the PR.

I'll finish updating the config file and then merge.

**Comment by djcunningham0** [comment]
Done and merged. Long live hackjohn!

**Pull request merged**
